### PR TITLE
Advanced pixel formats.

### PR DESCRIPTION
**Pull request by iabdalkader**
* Use a new pixel format that encodes flags, id and bpp in the pixel format itself.

Start with imlib.h to understand what's going on.

### CODE REVIEW

File: src/omv/imlib/imlib.c [hunk 1]
```diff
@@ -255,40 +255,39 @@ void image_copy(image_t *dst, image_t *src)
...
-        case IMAGE_BPP_RGB565: {
+        case PIXFORMAT_RGB565: {
             return IMAGE_RGB565_LINE_LEN_BYTES(ptr) * ptr->h;
         }
-        case IMAGE_BPP_BAYER: {
+        case PIXFORMAT_BAYER: {
             return ptr->w * ptr->h;
         }
-        default: { // JPEG
+        case PIXFORMAT_JPEG: {
```
Comment (kwagyeman) [inline src/omv/imlib/imlib.c:271]: Please check if this will actually work.

File: src/omv/ports/stm32/sensor.c [hunk 2]
```diff
@@ -956,36 +957,38 @@ int sensor_snapshot(sensor_t *sensor, image_t *image, uint32_t flags)
...
-            MAIN_FB()->bpp = IMAGE_BPP_BAYER;
+            MAIN_FB()->pixfmt = PIXFORMAT_BAYER;
+            MAIN_FB()->pixfmt_id = sensor->hw_flags.bayer;
             #if (OMV_ENABLE_SENSOR_MDMA == 1)
             // Flush data for MDMA
             SCB_InvalidateDCache_by_Addr(buffer->data, w * h);
             #endif
             break;
-        case PIXFORMAT_JPEG:
+        case PIXFORMAT_JPEG: {
```
Comment (kwagyeman) [inline src/omv/ports/stm32/sensor.c:982]: Please check this will actually work.

File: src/omv/imlib/framebuffer.c [hunk 3]
```diff
@@ -125,8 +127,9 @@ void framebuffer_update_jpeg_buffer()
     framebuffer_initialize_image(&main_fb_src);
     image_t *src = &main_fb_src;
 
-    if (framebuffer->streaming_enabled && jpeg_framebuffer->enabled) {
-        if (src->bpp > 3) {
+    if (src->pixfmt != PIXFORMAT_INVALID &&
+            framebuffer->streaming_enabled && jpeg_framebuffer->enabled) {
+        if (src->pixfmt == PIXFORMAT_JPEG) {
```
Comment (kwagyeman) [inline src/omv/imlib/framebuffer.c:132]: Not sure if this will work.

**Review by kwagyeman — APPROVED** [review]
Per your own comments about big PRs I can't review this. It's just too much.

However, the only thing I would check is all code dealing with jpegs since pixformat is not a static value anymore.

**Comment by iabdalkader** [comment]
> Per your own comments about big PRs I can't review this. It's just too much.

That's true, but this has to be done all at once because it changes the pixel format, on the other hand I'd argue that #1445 (which also edits 15 files) can easily be reduced to more reviewable PRs.

> However, the only thing I would check is all code dealing with jpegs since pixformat is not a static value anymore.

You're right, the bpp can be different for images and fb but not the sensor->pixformat. That's why I wanted to use the format id thing but not enough bits to implement it.

**Review by iabdalkader — COMMENTED** [review]
Reviewed it again.

File: src/omv/imlib/imlib.h [hunk 4]
```diff
@@ -336,75 +336,134 @@ extern const uint16_t ironbow_table[256];
...
+    PIXFORMAT_BPP_BINARY    = 0,
+    PIXFORMAT_BPP_GRAYSCALE = 1,
+    PIXFORMAT_BPP_RGB565    = 2,
+    PIXFORMAT_BPP_YUV422    = 2,
+    PIXFORMAT_BPP_BAYER     = 1,
+} pixformat_bpp_t;
+
+// Pixel format IDs.
+typedef enum {
+  PIXFORMAT_ID_BAYER_BGGR   = 0, // < Note: about bayer sub-formats >
```
Comment (kwagyeman) [inline src/omv/imlib/imlib.h:350]: I may add more bayer formats to support automotive sensors that have clear filters...

File: src/omv/imlib/imlib.h [hunk 5]
```diff
@@ -336,75 +336,134 @@ extern const uint16_t ironbow_table[256];
...
 
 typedef struct image {
-    int w;
-    int h;
-    int bpp;
+    uint32_t w;
+    uint32_t h;
+    PIXFORMAT_STRUCT;
     union {
         uint8_t *pixels;
```
Comment (kwagyeman) [inline src/omv/imlib/imlib.h:452]: Can you do another PR to just pick pixels/data as the pointer versus having both?

File: src/omv/imlib/imlib.h [hunk 6]
```diff
@@ -336,75 +336,134 @@ extern const uint16_t ironbow_table[256];
...
-    __typeof__ (image) _image = (image); \
-    (_image->bpp == IMAGE_BPP_RGB565) || \
-    (_image->bpp == IMAGE_BPP_BAYER) || \
-    (_image->bpp >= IMAGE_BPP_JPEG); \
-})
+#define IMAGE_IS_MUTABLE(image)             (image->is_mutable)
+#define IMAGE_IS_COLOR(image)               (image->is_color)
+#define IMAGE_IS_MUTABLE_BAYER(image)       (image->is_mutable || image->is_bayer)
+// TODO: What's this needed for ? This is true for any image.
+#define IMAGE_IS_MUTABLE_BAYER_JPEG(image)  (image->is_mutable || image->is_bayer || image->is_compressed)
```
Comment (kwagyeman) [inline src/omv/imlib/imlib.h:466]: It's basically is valid. So, just replace this with is valid.

File: src/omv/modules/py_image.c [hunk 7]
```diff
@@ -358,129 +358,112 @@ STATIC mp_obj_t py_image_getiter(mp_obj_t o_in, mp_obj_iter_buf_t *iter_buf)
...
-        }
+    image_t *image = py_image_cobj(self);
+    if (image->pixfmt == PIXFORMAT_JPEG
+            && image->pixels[0] == 0xFE
+            && image->pixels[image->size-1] == 0xFE) {
+        // print for ide.
+        print->print_strn(print->data, (const char *) image->pixels, image->bpp);
+    } else {
+        mp_printf(print,
+                "width: %u height: %u size:%u pixfmt=\"%s\" is_color: %u is_mutable: %u is_bayer %u is_compressed: %u",
```
Comment (kwagyeman) [inline src/omv/modules/py_image.c:371]: Please keep json object notation above.

Comment (kwagyeman) [reply src/omv/modules/py_image.c:371]: E.g. add the {} and escapes.

Comment (kwagyeman) [reply src/omv/modules/py_image.c:371]: Also, folks might depend on the text being the same... so, I would not change the string output.

File: src/omv/modules/py_image.c [hunk 8]
```diff
@@ -578,20 +561,22 @@ static mp_obj_t py_image_subscr(mp_obj_t self_in, mp_obj_t index, mp_obj_t value
...
                     }
                     return mp_const_none;
                 }
-                mp_uint_t i = mp_get_index(self->base.type, self->_cobj.bpp, index, false);
-                self->_cobj.data[i] = mp_obj_get_int(value);
+                mp_uint_t i = mp_get_index(self->base.type, image->bpp, index, false);
+                image->data[i] = mp_obj_get_int(value);
                 return mp_const_none;
             }
+            default:
```
Comment (kwagyeman) [inline src/omv/modules/py_image.c:578]: Braces

File: src/omv/modules/py_imageio.c [hunk 9]
```diff
@@ -186,7 +186,10 @@ mp_obj_t py_imageio_read(uint n_args, const mp_obj_t *args, mp_map_t *kw_args)
         stream->ms = ms;
         read_long(fp, (uint32_t *) &image.w);
         read_long(fp, (uint32_t *) &image.h);
-        read_long(fp, (uint32_t *) &image.bpp);
+        uint32_t bpp;
+        read_long(fp, (uint32_t *) &bpp);
+        // TODO: workaround for backwards compatibility until the stream version is updated.
+        image.pixfmt = (bpp == 1) ? PIXFORMAT_GRAYSCALE : PIXFORMAT_RGB565;
```
Comment (kwagyeman) [inline src/omv/modules/py_imageio.c:192]: It could be 0, 1, 2, 3, or Jpeg...

File: src/omv/modules/py_imageio.c [hunk 10]
```diff
@@ -351,8 +354,9 @@ mp_obj_t py_imageio_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_
         stream->f_count = mp_obj_get_int(args[1]);
         stream->w       = mp_obj_get_int(image_info[0]);
         stream->h       = mp_obj_get_int(image_info[1]);
-        stream->bpp     = mp_obj_get_int(image_info[2]);
-        image_t image = {.w = stream->w, .h = stream->h, .bpp = stream->bpp, .pixels = NULL};
+        // TODO: workaround for backwards compatibility until the stream version is updated.
```
Comment (kwagyeman) [inline src/omv/modules/py_imageio.c:357]: It could be 0, 1, 2, 3 or jpeg.

**Review by kwagyeman — COMMENTED** [review]
You should check the code for software jpeg. I remember looking at that code and noticing that it used 1/2 for grayscale/rgb liberally.

File: src/omv/imlib/imlib.h (hunk 5 above)
Comment (iabdalkader) [reply src/omv/imlib/imlib.h:452]: Yes, open an issue.

File: src/omv/imlib/imlib.h (hunk 6 above)
Comment (iabdalkader) [reply src/omv/imlib/imlib.h:466]: Removed.

File: src/omv/modules/py_image.c (hunk 7 above)
Comment (iabdalkader) [reply src/omv/modules/py_image.c:371]: Reverted to old print.

File: src/omv/modules/py_image.c (hunk 8 above)
Comment (iabdalkader) [reply src/omv/modules/py_image.c:578]: Not needed unless you declare variables.

File: src/omv/imlib/imlib.h (hunk 4 above)
Comment (iabdalkader) [reply src/omv/imlib/imlib.h:350]: Wouldn't that be grayscale ? Anyway so I switched back to the original format split:

```c
// 31......28  27  26  25  24  23..........16  15...........8  7.............0
// <RESERVED>  MF  CF  JF  RF  <PIXFORMAT_ID>  <SUBFORMAT_ID>  <BYTES_PER_PIX>
// NOTE: Bit 31-30 must Not be used for pixformat_t to be used as mp_int_t.
```
And increased `sensor.hw_flags.bayer` to 3 bits (8 possible sub-formats), you can add more if needed. You can also test if an image is *any* bayer by using `image->pixfmt_id == PIXFORMAT_ID_BAYER` and you can either switch on `img->pixfmt`, `img->pixfmt_id` or `img->subfmt_id` for subformats.

File: src/omv/modules/py_imageio.c (hunk 9 above)
Comment (iabdalkader) [reply src/omv/modules/py_imageio.c:192]: How do I know if it's bayer or grayscale ? The stream file format must be updated. That's why I'm doing this, everything was built on the assumption that BPP is unique.

**Comment by iabdalkader** [comment]
See https://github.com/openmv/openmv/pull/1448/commits/e223762664eb6050597137de1ebe52e393bd0474 for updates.

ImageIO can be fixed separately, but it needs to use a pixel format in both memory and file streams. I could try to maintain backwards compatibility by incrementing the version.

**Comment by kwagyeman** [comment]
Okay, it's fine to merge.
